### ISSUE TITLE
Load subdetector selectively

### ISSUE THREAD

**Issue by TrT-TOT** [AIDASoft/DD4hep#688]
Hello, I have a commen question regardless of the version of DD4hep.
  The situation is I want to implement a commonly used function in DD4hep, loading different subdetectors selectively. It's very easy to do that with geant4, but I have no idea how to do that in DD4hep. Could you please give some advice on it?

Best regards!

**Comment by MarkusFrankATcernch** [comment]
Hi,

could you please be a bit more specific?
Do you mean how to ignore and not load certain subdetectors?

Markus

**Comment by TrT-TOT** [comment]
Hi, Frank,
  Yes, you are right. The struture of my geaometry files are arranged as:
compact.xml
-subdetector1.xml
-subdetector2.xml
-...

All the subdetector's xmls are included in the compact.xml to load them. Is there a  convenient way to manage them without changing compact.xml? Or is there a better solution to ignore and not load certain subdetectors?

**Comment by TrT-TOT** [comment]
I don't know whether you have any comment, but I've thought of a solution via Geant4Converter in DDG4. Thank you anyway.

**Comment by MarkusFrankATcernch** [comment]
Hi TrT-TOT
I have currently a lousy internet connection. I shall answer you in more detail next week when I am back in the office.
Is this acceptable?
Cheers, Markus

**Comment by TrT-TOT** [comment]
Sure, please feel free to not answer me. We are not stuck here. Let's make DD4hep a better framework together

**Comment by MarkusFrankATcernch** [comment]
So....
In pure dd4hep the detector is build in three steps
1) the detector is intialized (for this you need the parameters of the world)
2) then one can add iteratively subdetectors
3) then the detector build ends, the geometry is closed, voxilization takes place etc. and the geometry and structural part of the detector description is usable.

then we come to the details. If you use the compact notation, you can do the same. 
All steps 1), 2), 3) are represented here by XML files you parse.

You have to ensure that in the xml read in step 1) the geometry is not closed.
In any xml corresponding to step 2) neither the geometry should be opened nor closed. 
consequently in the xml representing step 3) the geometry should be closed.

Is this what you had in mind?
given that you can have <include/> directives at multiple levels I personally do not see a big advantage of not having a primary xml.
However, I know that e.g. the FCC guys do something I described above.

Cheers,
Markus

**Comment by TrT-TOT** [comment]
This may take me some time to digest your idea. But I think our idea is similar intuitively. I'll not reply you soon. Please forgive me for that.

**Comment by MarkusFrankATcernch** [comment]
I added an example in <DD4hep>/examples/CLICSid on how to selectively load xml files from the command line using the compact notation.

watch out for for the test:  t_CLICSiD_multiple_inputs

**Comment by TrT-TOT** [comment]
Hi Markus,

  I have to admit that I'm a noob in coding. Your new example is concise and intuitive, and I want to share my approach with you, to see if there is any problem.

  Our geometry conversation service is implemented from`DDG4/plugins/Geant4DetectorSensitivesConstruction.cpp`. The following code snippet shows how we load geometry:

    dd4hep::Detector& description = dd4hep::Detector::getInstance();
    description.fromCompact("file:"+ m_Path_to_XML_FILE);
    DetElement world = description.world();
    Geant4Converter conv(description);
    Geant4GeometryInfo *geo = conv.create(world).detach();
    G4VPhysicalVolume *physWorld = geo->world();
    return physWorld;
    
  To load multi detectors, we change it into:

    dd4hep::Detector& description = dd4hep::Detector::getInstance();
    description.fromCompact("file:"+ m_Path_to_main_XML_FILE);
    description.fromCompact("file:"+ m_Path_to_sub_XML_FILE_1);
    description.fromCompact("file:"+ m_Path_to_sub_XML_FILE_2);
    ...

  I don't fully understand why but it really works... According to your newly added example, I guess what I do is similar to `geoDisplay -input 1.XML -input 2.XML`? Besides, I noticed that there is an additional`DDG4/plugins/GeantDetectorSensitivesConstruction.cpp`, but we don't write  the`ConstructSDandField()` function in our code. Will this lead to a loss of some detector information, like sensitive detector?

PS: I really learned a lot from your code, your example as well as FCCSW and we'll optimize our code accordingly!

**Comment by MarkusFrankATcernch** [comment]
This looks correct.

* The Geant4 code is essentially 1:1 present in Geant4DetectorGeometryConstruction::constructGeo.
* For the second part this is exactly what geoDisplay does internally given multiple "-input" arguments.
* The important part is in the xml:
```
  <geometry open="false" close="false"/>
```
These you may not forget to add accordingly.

**Comment by TrT-TOT** [comment]
Thank you for your reply. I think we can close this issue now.
